src/utilities.jl

```diff
@@ -0,0 +1,42 @@
+# high-level utilities
+
+macro collect(ex)
+    quote
+        isactive() || @warn("No ITT collector present. Make sure you have a collector attached to your process, e.g., by running under VTune.")
+        resume()
+        # Use Expr(:tryfinally, ...) so we don't introduce a new soft scope.
+        # TODO: switch to solution once https://github.com/JuliaLang/julia/pull/39217 is resolved
+        $(Expr(:tryfinally, esc(ex), :(pause())))
+    end
+end
+
+macro task(ex...)
+    length(ex) >= 2 || error("Usage: IntelITT.@task [domain=...] name::String code")
+    name = ex[end-1]
+    code = ex[end]
+    kwargs = ex[1:end-2]
+
+    # handle kwargs
+    domain = nothing
+    for kwarg in kwargs
+        key::Symbol, val = kwarg.args
+        if key === :domain
+            domain = val
+        else
+            error("unknown keyword argument: $key")
+        end
+    end
+    if domain === nothing
+        domain = :(Domain(string($__module__)))
```

Comment:
This will call `Domain("modulename")` at each invocation, which is probably not ideal.

In NVTX.jl, I define a hidden `const` in each module:
https://github.com/JuliaGPU/NVTX.jl/blob/d5570780ab22569659925742d4257e9e131fec50/src/macro.jl#L7-L14
evaluate it within the macro (i.e. `$(Domain(__module__))`), and lazily initialize it when first used.

Reply:
It relies on pointers being serialized to `C_NULL` (I'm not sure if this is actually documented anywhere?), and then the macros will lazily initialize:
https://github.com/JuliaGPU/NVTX.jl/blob/d5570780ab22569659925742d4257e9e131fec50/src/macro.jl#L74

Reply:
> It relies on pointers being serialized to `C_NULL`

Ah, of course, that's clever. I'm not sure it's worth relying on this idiosyncrasy though; creating a domain costs 20ns when not being profiled, and 30ns under VTune. I'd assume tasks are generally used for more coarse-grained operations, or what is your experience?